journal/src/main/java/io/zeebe/journal/file/record/SBESerializer.java

```diff
@@ -0,0 +1,137 @@
+import io.zeebe.journal.file.RecordMetadataEncoder;
+import java.nio.BufferOverflowException;
+import org.agrona.DirectBuffer;
+import org.agrona.MutableDirectBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+
+/** The serializer that writes and reads a journal record according to the SBE schema defined. */
+public final class SBESerializer implements JournalRecordSerializer {
+
+  private final MessageHeaderEncoder headerEncoder = new MessageHeaderEncoder();
+  private final RecordMetadataEncoder metadataEncoder = new RecordMetadataEncoder();
+  private final RecordDataEncoder recordEncoder = new RecordDataEncoder();
+
+  private final MessageHeaderDecoder headerDecoder = new MessageHeaderDecoder();
+  private final RecordMetadataDecoder metadataDecoder = new RecordMetadataDecoder();
+  private final RecordDataDecoder recordDecoder = new RecordDataDecoder();
+
+  @Override
+  public int writeData(
+      final RecordData record, final MutableDirectBuffer buffer, final int offset) {
+    if (offset + getSerializedLength(record) > buffer.capacity()) {
+      throw new BufferOverflowException();
+    }
+
+    headerEncoder
+        .wrap(buffer, offset)
+        .blockLength(recordEncoder.sbeBlockLength())
+        .templateId(recordEncoder.sbeTemplateId())
+        .schemaId(recordEncoder.sbeSchemaId())
+        .version(recordEncoder.sbeSchemaVersion());
+
+    recordEncoder.wrap(buffer, offset + headerEncoder.encodedLength());
+
+    recordEncoder
+        .index(record.index())
+        .asqn(record.asqn())
+        .putData(record.data(), 0, record.data().capacity());
+
+    return headerEncoder.encodedLength() + recordEncoder.encodedLength();
+  }
+
+  @Override
+  public int writeMetadata(
+      final RecordMetadata metadata, final MutableDirectBuffer buffer, final int offset) {
+
+    headerEncoder
+        .wrap(buffer, offset)
+        .blockLength(metadataEncoder.sbeBlockLength())
+        .templateId(metadataEncoder.sbeTemplateId())
+        .schemaId(metadataEncoder.sbeSchemaId())
+        .version(metadataEncoder.sbeSchemaVersion());
+
+    metadataEncoder.wrap(buffer, offset + headerEncoder.encodedLength());
+
+    metadataEncoder.checksum(metadata.checksum()).length(metadata.length());
+
+    return headerEncoder.encodedLength() + metadataEncoder.encodedLength();
+  }
+
+  @Override
+  public int getMetadataLength() {
+    return headerEncoder.encodedLength() + metadataEncoder.sbeBlockLength();
+  }
+
+  @Override
+  public boolean hasMetadata(final DirectBuffer buffer, final int offset) {
+    headerDecoder.wrap(buffer, offset);
+    return (headerDecoder.schemaId() == metadataDecoder.sbeSchemaId()
+        && headerDecoder.templateId() == metadataDecoder.sbeTemplateId());
+  }
+
+  @Override
+  public RecordMetadata readMetadata(final DirectBuffer buffer, final int offset) {
+    if (!hasMetadata(buffer, offset)) {
+      throw new InvalidRecordException("Cannot read buffer. Header does not match.");
```

Comment:
This isn't being handled right? I guess the correct behavior (for now) is to assume that the log has reached the end and this data is truncated garbage

Reply:
We are now always zeroing out then next 8 bytes, after each record. So if the log is truncated, then we expect that `hasMetadata` returns false. The reader is expected to call `hasMetadata` before calling this method. See `JournalRecordReaderUtil`. If there is not metadata, it assumes it has reached the end. So ideally this exception should never happen, except if the log is corrupted.
Would you suggest to ignore the exception in the reader, and treat it as end of the log?